File: backend/app/schemas/talent_list.py
import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from app.models.talent_list import TalentListMemberStatus, TalentListSource
from app.schemas.candidate import CandidateResponse
from app.schemas.evaluation import ProjectCandidateResponse


# ── Lists ─────────────────────────────────────────────────────────────────────


class TalentListCreate(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    project_id: uuid.UUID | None = None
    filters_json: dict | None = None
    source: TalentListSource = TalentListSource.MANUAL
    # Optional initial members. If omitted, an empty list is created.
    candidate_ids: list[uuid.UUID] = Field(default_factory=list)


class TalentListUpdate(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=255)
    # Pass project_id=null explicitly to unbind a list. Pydantic's exclude_unset
    # is what the service uses to distinguish "unset" from "set to null".
    project_id: uuid.UUID | None = None
    filters_json: dict | None = None


class TalentListSummaryResponse(BaseModel):
    """Lightweight list summary used in the index page (no members)."""
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    name: str
    project_id: uuid.UUID | None
    project_name: str | None = None
    client_name: str | None = None
    filters_json: dict | None
    source: TalentListSource
    member_count: int = 0
    # Outreach progress breakdown so the sidebar/index page can show
    # "12 人 · 已联系 4 · 已回复 2" without re-fetching every list.
    status_counts: dict[str, int] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class TalentListMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    list_id: uuid.UUID
    candidate_id: uuid.UUID
    status: TalentListMemberStatus
    hunter_note: str | None
    added_at: datetime
    updated_at: datetime
    candidate: CandidateResponse | None = None
    # Project-level evaluation joined in when the parent list is bound to a
    # project. Null otherwise. Once a member is promoted ("已推进"), this
    # carries the EA match_score, recommendation, project_candidate.status,
    # etc., so the list page can render the full follow-up state.
    project_evaluation: ProjectCandidateResponse | None = None


class TalentListDetailResponse(TalentListSummaryResponse):
    members: list[TalentListMemberResponse] = Field(default_factory=list)


# ── Members ───────────────────────────────────────────────────────────────────


class TalentListMembersAdd(BaseModel):
    candidate_ids: list[uuid.UUID] = Field(..., min_length=1)


class TalentListMemberUpdate(BaseModel):
    status: TalentListMemberStatus | None = None
    hunter_note: str | None = None

File: frontend/src/types/talentList.ts
import type { CandidateResponse } from './candidate'
import type { ProjectCandidateResponse } from './evaluation'

export type TalentListSource = 'manual' | 'ca_conversation'

export type TalentListMemberStatus =
  | 'not_contacted'
  | 'contacted'
  | 'scheduled'
  | 'declined'
  | 'added_to_project'

/**
 * Filters payload saved alongside a list. Mirrors a subset of
 * CandidateSearchRequest so the list can be "re-run" later. We store this as
 * an opaque dict server-side; the frontend interprets it.
 */
export interface TalentListFiltersJson {
  query?: string | null
  location?: string | null
  min_years_experience?: number | null
  max_years_experience?: number | null
  current_company?: string | null
  schools?: string[] | null
  exclude_companies?: string[] | null
  exclude_locations?: string[] | null
  exclude_query?: string | null
  // Free-form: anything we want to remember about the search context.
  [key: string]: unknown
}

export interface TalentListSummaryResponse {
  id: string
  name: string
  project_id: string | null
  project_name: string | null
  client_name: string | null
  filters_json: TalentListFiltersJson | null
  source: TalentListSource
  member_count: number
  status_counts: Partial<Record<TalentListMemberStatus, number>>
  created_at: string
  updated_at: string
}

export interface TalentListMemberResponse {
  id: string
  list_id: string
  candidate_id: string
  status: TalentListMemberStatus
  hunter_note: string | null
  added_at: string
  updated_at: string
  candidate: CandidateResponse | null
  /**
   * Project-level evaluation joined in by the backend when the parent list is
   * bound to a project. Null when the list is unassigned, or when this member
   * has not been promoted to project evaluation yet.
   *
   * Once present, the recruiter UI uses this to:
   * - Render the "评估中 / 已评估 N 分 / 评估失败" variants of the 已推进 badge
   * - Open the EvaluationReportDrawer to view match_score, recommendation, etc.
   * - Edit project_candidates.status (recommended / interviewed / eliminated)
   */
  project_evaluation: ProjectCandidateResponse | null
}

export interface TalentListDetailResponse extends TalentListSummaryResponse {
  members: TalentListMemberResponse[]
}

export interface TalentListCreate {
  name: string
  project_id?: string | null
  filters_json?: TalentListFiltersJson | null
  source?: TalentListSource
  candidate_ids?: string[]
}

export interface TalentListUpdate {
  name?: string
  project_id?: string | null
  filters_json?: TalentListFiltersJson | null
}

export interface TalentListMemberUpdate {
  status?: TalentListMemberStatus
  hunter_note?: string | null
}
